integrations/unftp-sbe/src/lib.rs

```diff
@@ -14,3 +14,221 @@
+    }
+}
+
+/// A wrapper around [`opendal::Metadata`] to implement [`libunftp::storage::Metadata`].
+pub struct OpendalMetadata(pub opendal::Metadata);
+
+impl storage::Metadata for OpendalMetadata {
+    fn len(&self) -> u64 {
+        self.0.content_length()
+    }
+
+    fn is_dir(&self) -> bool {
+        self.0.is_dir()
+    }
+
+    fn is_file(&self) -> bool {
+        self.0.is_file()
+    }
+
+    fn is_symlink(&self) -> bool {
+        false
+    }
+
+    fn modified(&self) -> storage::Result<std::time::SystemTime> {
+        self.0.last_modified().map(Into::into).ok_or_else(|| {
+            storage::Error::new(storage::ErrorKind::LocalError, "no last modified time")
+        })
+    }
+
+    fn gid(&self) -> u32 {
+        0
+    }
+
+    fn uid(&self) -> u32 {
+        0
+    }
+}
+
+fn cvt_err(err: opendal::Error) -> storage::Error {
+    let kind = match err.kind() {
+        opendal::ErrorKind::NotFound => storage::ErrorKind::PermanentFileNotAvailable,
+        opendal::ErrorKind::AlreadyExists => storage::ErrorKind::PermanentFileNotAvailable,
+        opendal::ErrorKind::PermissionDenied => storage::ErrorKind::PermissionDenied,
+        _ => storage::ErrorKind::LocalError,
+    };
+    storage::Error::new(kind, err)
+}
+
+fn cvt_path(path: &Path) -> storage::Result<&str> {
+    path.to_str().ok_or_else(|| {
+        storage::Error::new(
+            storage::ErrorKind::LocalError,
+            "Path is not a valid UTF-8 string",
+        )
+    })
+}
+
+#[async_trait::async_trait]
+impl<User: UserDetail> StorageBackend<User> for OpendalStorage {
+    type Metadata = OpendalMetadata;
+
+    async fn metadata<P: AsRef<Path> + Send + Debug>(
+        &self,
+        _: &User,
+        path: P,
+    ) -> storage::Result<Self::Metadata> {
+        let metadata = self
+            .op
+            .stat(cvt_path(path.as_ref())?)
+            .await
+            .map_err(cvt_err)?;
+        Ok(OpendalMetadata(metadata))
+    }
+
+    async fn list<P: AsRef<Path> + Send + Debug>(
+        &self,
+        _: &User,
+        path: P,
+    ) -> storage::Result<Vec<storage::Fileinfo<PathBuf, Self::Metadata>>>
+    where
+        Self::Metadata: storage::Metadata,
+    {
+        let ret = self
+            .op
+            .list(cvt_path(path.as_ref())?)
+            .await
+            .map_err(cvt_err)?
+            .into_iter()
+            .map(|x| {
+                let (path, metadata) = x.into_parts();
+                storage::Fileinfo {
+                    path: path.into(),
+                    metadata: OpendalMetadata(metadata),
+                }
+            })
+            .collect();
+        Ok(ret)
+    }
+
+    async fn get<P: AsRef<Path> + Send + Debug>(
+        &self,
+        _: &User,
+        path: P,
+        start_pos: u64,
+    ) -> storage::Result<Box<dyn tokio::io::AsyncRead + Send + Sync + Unpin>> {
+        let buf = self
+            .op
+            .read_with(cvt_path(path.as_ref())?)
```

Comment:
Please use `op.reader()` and `into_futures_async_read` instead of reading all data in memory.